sync/src/block/extension.rs

```diff
@@ -138,34 +196,29 @@ impl Extension {
         }
     }
 
-    fn initial_state(client: Arc<Client>, snapshot_target: Option<(BlockHash, u64)>) -> State {
-        let (hash, num) = match snapshot_target {
-            Some(target) => target,
-            None => return State::Full,
-        };
-        let header = match client.block_header(&num.into()) {
-            Some(ref h) if h.hash() == hash => h.clone(),
-            _ => return State::SnapshotHeader(hash, num),
-        };
-        if client.block_body(&hash.into()).is_none() {
-            let parent_hash = header.parent_hash();
-            let parent =
-                client.block_header(&parent_hash.into()).expect("Parent header of the snapshot header must exist");
-            return State::SnapshotBody {
-                header,
-                prev_root: parent.transactions_root(),
+    fn move_state(&mut self) {
+        let next_state = self.state.next(&self.client);
+        cdebug!(SYNC, "Transitioning the state to {:?}", next_state);
+        if discriminant(&next_state) == discriminant(&State::Full) {
+            let best_hash = match &self.state {
+                State::SnapshotHeader(hash, _) => *hash,
+                State::SnapshotBody {
+                    header,
+                    ..
+                } => header.hash(),
+                State::SnapshotChunk {
+                    block,
+                    ..
+                } => *block,
+                State::Full => unreachable!("Trying to transition state from State::Full"),
```

Comment:
`unreachable` seems to be used as a debug designator in a pretty local scope. 
https://doc.rust-lang.org/std/macro.unreachable.html
`panic` would be more precise if you meant to throw an error on the violation of interface assumption.

Reply:
Fixed.